### PR TITLE
 Implement traffic redirection exclusion based on proxy config and user-provided values

### PR DESCRIPTION
**Pull request by ishustava**
* If `-proxy-id` is provided to the redirect-traffic command, exclude any listener ports from inbound traffic redirection. This includes `envoy_prometheus_bind_addr`, `envoy_stats_bind_addr`, and the `ListenerPort` from the `Expose` configuration.
* Allow users to provide additional inbound and outbound ports, outbound CIDRs and additional user IDs to be excluded from traffic redirection. This affects both the traffic-redirect command and the iptables SDK package.
* Additionally, use the proxy outbound port from the proxy config if it's set.

### CODE REVIEW

File: command/connect/redirecttraffic/redirect_traffic.go
Comment (freddygv) [inline command/connect/redirecttraffic/redirect_traffic.go:63]: Would it be valuable to also be able to exclude GIDs?

File: command/connect/redirecttraffic/redirect_traffic.go
Comment (freddygv) [inline command/connect/redirecttraffic/redirect_traffic.go:166]: TransparentProxy should have a nil check here since it's a pointer. 

Consul currently always returns a non-nil struct even if it's empty, but that's likely going to change for GA.

File: command/connect/redirecttraffic/redirect_traffic.go
Comment (freddygv) [inline command/connect/redirecttraffic/redirect_traffic.go:191]: One thing I'm realizing is missing is this bit of magic:
https://www.consul.io/docs/connect/registration/service-registration#checks

If someone enables this, Consul will dynamically expose checks on a listener for each check:
https://github.com/hashicorp/consul/blob/c343544fef094a505d22cf60bd79635cf8e2553e/agent/agent.go#L3839

We don't expose what port is allocated in any of our APIs, similarly to how we don't expose check definitions via any of our APIs, only check statuses. See this comment for more information on why: https://github.com/hashicorp/consul/issues/7330#issuecomment-592015038

There are a few potential solutions to discover those ports off the top of my head:

1. Expand the `structs.HealthCheck` type to include the exposed path.
2. Add a Meta map to checks, like we have for services, and include the exposed path there.
3. Include all check definition details in `structs.HealthCheck`. That comes with the security risk mentioned above.

File: command/connect/redirecttraffic/redirect_traffic.go
Comment (freddygv) [inline command/connect/redirecttraffic/redirect_traffic.go:196]: What if the defaulting is pulled out the else? These lines are really far removed from the `if` condition so it's not obvious what case this is. 

```suggestion
	cfg := iptables.Config{
		ProxyUserID: c.proxyUID
		ProxyInboundPort: c.proxyInboundPort
		ProxyOutboundPort: c.proxyOutboundPort
	}
	
	if c.proxyID != "" {
		.....
	}
```

**Review by freddygv — COMMENTED** [review]
Thanks for expanding this! I made a couple minor comments and one less trivial one about the Expose config. Depending on timing we could push the solution to that issue to after beta2 though.

File: command/connect/redirecttraffic/redirect_traffic.go
Comment (freddygv) [reply command/connect/redirecttraffic/redirect_traffic.go:191]: So based on the discussion today we should go with #1.

The exposed ports are generated in a few places. 

One is where we add checks:
https://github.com/hashicorp/consul/blob/0c9555ec0e2366996cff09d39d1bcb40eaeee435/agent/agent.go#L2610
https://github.com/hashicorp/consul/blob/0c9555ec0e2366996cff09d39d1bcb40eaeee435/agent/agent.go#L2541

In `addCheck` it's easy to expand the HealthCheck type, since we already have the `check` pointer to it. 

And the other is where we add services with checks:
https://github.com/hashicorp/consul/blob/0c9555ec0e2366996cff09d39d1bcb40eaeee435/agent/agent.go#L3791
https://github.com/hashicorp/consul/blob/0c9555ec0e2366996cff09d39d1bcb40eaeee435/agent/agent.go#L3818

In these two other methods we'll need to query `a.State.Check`, I think, and modify that.

File: command/connect/redirecttraffic/redirect_traffic.go
Comment (ishustava) [reply command/connect/redirecttraffic/redirect_traffic.go:63]: I think it could be. The main reason we're adding this is to better support the VM use case when you might want to exclude other processes/uids/gids from traffic redirection. I'm inclined to wait for users' feedback on this one because I'm not sure what would be valuable there.

File: command/connect/redirecttraffic/redirect_traffic.go
Comment (ishustava) [reply command/connect/redirecttraffic/redirect_traffic.go:166]: Great catch! let me update it.

File: command/connect/redirecttraffic/redirect_traffic.go
Comment (ishustava) [reply command/connect/redirecttraffic/redirect_traffic.go:191]: That's a great find! Let me address this in a separate PR so that this one can be included in beta2.

File: command/connect/redirecttraffic/redirect_traffic.go
Comment (ishustava) [reply command/connect/redirecttraffic/redirect_traffic.go]: done!

**Review by freddygv — APPROVED** [review]
:shipit:

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/359312.

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit 8dffb8913105572af71f582bd204c6f473d4a915 onto `release/1.10.x` succeeded!

**Pull request merged**
